File: Minix-Platform-Loader/src/main/java/dev/racci/minix/platforms/PlatformProvider.java
package dev.racci.minix.platforms;

import java.io.*;
import java.net.*;
import java.util.*;
import java.util.function.*;

interface PlatformProvider extends Function<File, URLClassLoader> {}

class PlatformProviderImpl implements PlatformProvider {
    private final Map<File, URLClassLoader> alreadyLoaded = new HashMap<>();

    @Override
    public URLClassLoader apply(File file) {
        URLClassLoader cached = alreadyLoaded.get(file);
        if(cached == null) {
            try {
                cached = new URLClassLoader(new URL[]{file.toURI().toURL()});
                alreadyLoaded.put(file, cached);
            } catch (MalformedURLException e) {
                e.printStackTrace();
            }
        }
        return cached;
    }
}

File: Minix-Platform-Loader/src/main/kotlin/dev/racci/minix/platforms/PlatformProvider.kt
package dev.racci.minix.platforms

import java.io.File
import java.net.MalformedURLException
import java.net.URLClassLoader
import java.util.function.Function
import java.util.logging.Level

internal interface PlatformProvider: Function<File, URLClassLoader>

internal class PlatformProviderImpl: PlatformProvider {

    private val alreadyLoaded: MutableMap<File, URLClassLoader> = HashMap()

    override fun apply(file: File): URLClassLoader {
        var cached = alreadyLoaded[file]
        if(cached == null) {
            try {
                cached = URLClassLoader(arrayOf(file.toURI().toURL()))
                alreadyLoaded[file] = cached
            } catch(e: MalformedURLException) {
                org.bukkit.plugin.PluginLogger.getAnonymousLogger().log(Level.FINEST, e.message)
            }
        }
        return cached!!
    }
}
